src/index/blockfilter.cpp

```diff
@@ -0,0 +1,289 @@
+
+    std::unique_ptr<CDBIterator> db_it(db.NewIterator());
+    db_it->Seek(std::make_pair(key_prefix, std::make_pair(DB_BLOCK_HEIGHT, start_height)));
+    for (int height = start_height; height <= stop_index->nHeight; ++height) {
+        std::pair<char, std::pair<char, int>> key, expected_key;
+        expected_key = std::make_pair(key_prefix, std::make_pair(DB_BLOCK_HEIGHT, height));
+
+        if (!db_it->Valid() || !db_it->GetKey(key) || key != expected_key) {
+            return false;
+        }
+
+        size_t i = height - start_height;
+        if (!db_it->GetValue(values[i])) {
+            return error("%s: unable to read value in %s at key (%c, (%c, %d))",
+                         __func__, index_name, key_prefix, DB_BLOCK_HEIGHT, height);
+        }
+
+        db_it->Next();
+    }
+
+    results.resize(stop_index->nHeight - start_height + 1);
+    for (const CBlockIndex* block_index = stop_index;
+         block_index && block_index->nHeight >= start_height;
+         block_index = block_index->pprev) {
+        uint256 block_hash = block_index->GetBlockHash();
+
+        size_t i = block_index->nHeight - start_height;
+        if (block_hash == values[i].first) {
+            results[i] = std::move(values[i].second);
+            continue;
+        }
+
+        std::pair<char, uint256> hash_key(DB_BLOCK_HASH, block_hash);
+        if (!db.Read(std::make_pair(key_prefix, hash_key), results[i])) {
+            return error("%s: unable to read value in %s at key (%c, (%c, %s))",
+                         __func__, index_name, key_prefix, DB_BLOCK_HASH, block_hash.ToString());
+        }
+    }
+
+    return true;
+}
+
+bool BlockFilterIndex::LookupFilter(const CBlockIndex* block_index, BlockFilter& filter_out) const
+{
+    std::vector<unsigned char> encoded_filter;
+    if (!LookupOne(*m_db, DB_FILTER, block_index, encoded_filter)) {
+        return false;
+    }
+
+    filter_out = BlockFilter(m_filter_type, block_index->GetBlockHash(), std::move(encoded_filter));
+    return true;
+}
+
+bool BlockFilterIndex::LookupFilterHeader(const CBlockIndex* block_index, uint256& header_out) const
+{
+    return LookupOne(*m_db, DB_FILTER_HEADER, block_index, header_out);
+}
+
+bool BlockFilterIndex::LookupFilterRange(int start_height, const CBlockIndex* stop_index,
+                                         std::vector<BlockFilter>& filters_out) const
+{
+    std::vector<std::vector<unsigned char>> encoded_filters;
+    if (!LookupRange(*m_db, m_name, DB_FILTER, start_height, stop_index, encoded_filters)) {
+        return false;
+    }
+
+    filters_out.resize(stop_index->nHeight - start_height + 1);
+
+    auto it = filters_out.rbegin();
+    auto encoded_filter_it = encoded_filters.rbegin();
+    const CBlockIndex* pindex = stop_index;
+
+    while (it != filters_out.rend()) {
+        *it = BlockFilter(m_filter_type, pindex->GetBlockHash(), std::move(*encoded_filter_it));
```

Comment:
Is assigning to the dereferenced pointer here instead of using a vector function an optimization? I can't see an obvious reason there is anything wrong with this, but just double checking.

Reply:
What do you mean by a vector function? This just seemed to be the most immediate way to do the assignment to me.

Reply:
I meant using something like `insert` instead of direct assignment. It's been a while since I've written a lot of c++.